tests/test_tomography_gate_set_tomography.py

```diff
@@ -0,0 +1,392 @@
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), S(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), S(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+            (qibo.symbols.Z(0), qibo.symbols.Z(1)),
+        ],
+    }
+    correct_observables[1] = [
+        SymbolicHamiltonian(h[0]).form for h in correct_observables[1]
+    ]
+    correct_observables[2] = [
+        SymbolicHamiltonian(reduce(lambda x, y: x * y, h)).form
+        for h in correct_observables[2]
+    ]
+    errors = {(0, 3): ValueError, (17, 1): IndexError}
+    if (j, nqubits) in [(0, 3), (17, 1)]:
+        with pytest.raises(errors[(j, nqubits)]):
+            prepared_observable = _get_observable(j, nqubits)
+    else:
+        prepared_observable = _get_observable(j, nqubits).form
+        groundtruth = correct_observables[nqubits][j]
+        assert groundtruth == prepared_observable
+
+
+def test_expectation_value(backend):
+    nqubits = 3
+    test_circuit = qibo.models.Circuit(nqubits)
+    test_circuit.add(gates.TOFFOLI(0, 1, 2))
+    k = 1
+    j = 1
+    with pytest.raises(ValueError):
+        expectation_val = _expectation_value(
+            test_circuit,
+            k,
+            j,
+            nshots=int(1e4),
+            backend=backend,
+        )
+
+
+def test_gate_tomography_invalid_qb_gate(backend):
+    nqubits = 3
+    test_circuit = qibo.models.Circuit(nqubits)
+    test_circuit.add(gates.TOFFOLI(0, 1, 2))
+    with pytest.raises(ValueError):
+        matrix_jk = _gate_tomography(
+            nqubits=nqubits,
+            gate=gates.TOFFOLI(0, 1, 2),
+            nshots=int(1e4),
+            invert_register=None,
+            noise_model=None,
+            backend=backend,
+        )
+
+
+def test_gate_tomography_mismatched_inputs(backend):
+    nqubits = 1
+    test_circuit = qibo.models.Circuit(nqubits)
+    test_circuit.add(gates.H(0))
+    test_gate = gates.CNOT(0, 1)
+    with pytest.raises(ValueError):
+        matrix_jk = _gate_tomography(
+            nqubits=nqubits,
+            gate=test_gate,
+            nshots=int(1e4),
+            invert_register=None,
+            noise_model=None,
+            backend=backend,
+        )
+
+
+def test_gate_tomography_with_invert_register(backend):
+    nqubits = 1
+    test_circuit = qibo.models.Circuit(nqubits)
+    test_circuit.add(gates.H(0))
+    matrix_jk = _gate_tomography(
+        nqubits=nqubits,
+        gate=None,
+        nshots=int(1e6),
+        invert_register=(0,),
+        noise_model=None,
+        backend=backend,
+    )
+    groundtruth = np.array([[1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0], [1, 1, 1, 1]])
```

Comment:
can you merge all the gate tomography tests, these three and all the others below, in a single (or at least a few) one using pytest parametrization as I did with the tests above?

Reply:
I don't understand how to merge these three into one test function. I'm testing for different aspects of the function.

Reply:
The only differences in the tests are the number of qubits, the gate and `invert_register`. Therefore, if you set  the three parameters `nqubits=range(1,4)`, `gate=[CNOT(0,1), TOFFOLI(0,1,2), None]`, `invert_register=[(0,), (1,), (0,1)]` not only you can merge everything to a single test, but you can also test all the combinations of the three parameters. In `pytest` you do that by appending these lines on top of the function
```
@pytest.mark.parametrize("nqubits", range(1,4))
@pytest.mark.parametrize("gate", [CNOT(0,1), TOFFOLI(0,1,2), None])
@pytest.mark.parametrize("invert_register", [(0,), (1,), (0,1)])
def test_gate_tomography(....):
```

Reply:
Ah ok I get what you mean, thanks.  I have some difficulty thinking about how to do the test for `invert_register=[(1,), (0,1)]`. Because `(1,)` and `(0,1)` are pertaining to 2 qubit circuits, the resulting matrix is 16 x 16. I can put in all 16 x 16 elements manually but it's only for a specific `gate`.  Could I trouble you to help with this?